lib/gears/shape.lua

```diff
@@ -168,6 +220,243 @@ function module.isosceles_triangle(cr, width, height)
+-- @tparam number corner_radius 
+function module.octogon(cr, width, height, corner_radius)
+    local corner_radius = corner_radius or math.min(10, math.min(width, height)/4)
+    local offset = math.sqrt( (corner_radius*corner_radius) / 2 )
+
+    cr:move_to(offset, 0)
+    cr:line_to(width-offset, 0)
+    cr:line_to(width, offset)
+    cr:line_to(width, height-offset)
+    cr:line_to(width-offset, height)
+    cr:line_to(offset, height)
+    cr:line_to(0, height-offset)
+    cr:line_to(0, offset)
+    cr:close_path()
+end
+
+--- A circle shape.
+--
+-- ![Shape example](@SHAPE_circle_SVG@)
+--
+-- @SHAPE_circle_EXAMPLE@
+--
+-- @param cr A cairo context
+-- @tparam number width The shape with
+-- @tparam number height The shape height
+function module.circle(cr, width, height)
+    local size = width < height and (width/2) or (height/2)
+    local w_offset, h_offset = (width - size*2)/2, (height - size*2)/2
+    cr:arc(w_offset + size, h_offset + size, size, 0, 2*math.pi)
+    cr:close_path()
+end
+
+--- A simple rectangle.
+--
+-- ![Shape example](@SHAPE_rectangle_SVG@)
+--
+-- @SHAPE_rectangle_EXAMPLE@
+--
+-- @param cr A cairo context
+-- @tparam number width The shape with
+-- @tparam number height The shape height
+function module.rectangle(cr, width, height)
+    cr:rectangle(0, 0, width, height)
+end
+
+--- A diagonal parallelogram with the bottom right corner at x=0 and top left
+-- at x=width.
+--
+-- ![Shape example](@SHAPE_parallelogram_SVG@)
+--
+-- @SHAPE_parallelogram_EXAMPLE@
+--
+-- @param cr A cairo context
+-- @tparam number width The shape with
+-- @tparam number height The shape height
+-- @tparam[opt=width/3] number thickness The parallelogram thickness
+function module.parallelogram(cr, width, height, thickness)
+    local thickness = thickness or width/3
+    cr:move_to(width-thickness, 0      )
+    cr:line_to(width          , 0      )
+    cr:line_to(thickness      , height )
+    cr:line_to(0              , height )
+    cr:close_path()
+end
+
+--- A losange
+-- @param cr A cairo context
+-- @tparam number width The shape with
+-- @tparam number height The shape height
+function module.losange(cr, width, height)
+    cr:move_to(width/2 , 0        )
+    cr:line_to(width   , height/2 )
+    cr:line_to(width/2 , height   )
+    cr:line_to(0       , height/2 )
+    cr:close_path()
+end
+
+--- A rounded rect with only some of the corners rounded.
+--
+-- ![Shape example](@SHAPE_partially_rounded_rect_SVG@)
+--
+-- @SHAPE_partially_rounded_rect_EXAMPLE@
+--
+-- @param cr A cairo context
+-- @tparam number width The shape with
+-- @tparam number height The shape height
+-- @tparam boolean tl If the top left corner is rounded
+-- @tparam boolean tr If the top right corner is rounded
+-- @tparam boolean br If the bottom right corner is rounded
+-- @tparam boolean bl If the bottom left corner is rounded
+-- @tparam number rad The corner radius
+function module.partially_rounded_rect(cr, width, height, tl, tr, br, bl, rad)
```

Comment:
Should `rounded_rect` be replaced with a call to this, just so that the code exists only once? (Dunno if that makes much of a difference)
Also, please move the two closer together.